### PR TITLE
Fixes #3492

### LINKED ISSUE #3492: a question about atom coordinates after embedding
**Configuration:**

- RDKit Version: 2020.03.6
- Operating system: linux
- Python version (if relevant): 3.8.3
- Are you using conda? yes
- If you are using conda, which channel did you install the rdkit from?   defaults/conda-forge
- If you are not using conda: how did you install the RDKit? 

**Description:**

Please see the two block of codes below, I believe they should have exactly same output, however only the first part can output coordinates of molecule, the second part will report something like:
****
Pre-condition Violation

Violation occurred on line 46 in file /home/conda/feedstock_root/build_artifacts/rdkit_1600678001365/work/Code/GraphMol/Conformer.cpp
Failed Expression: dp_mol->getNumAtoms() == d_positions.size()
****

Please note that I did not run them continuously, instead I comment the second part when I run the first part and vice versa

```
import rdkit
from rdkit import Chem
from rdkit.Chem import AllChem
import numpy as np
import copy


print(rdkit.__version__)

s = 'CC'

m = Chem.MolFromSmiles(s)
m = Chem.AddHs(m)
AllChem.EmbedMolecule(m)
print(m.GetConformers()[0].GetPositions())

#######################################################################
def read_smile(s):
    m = Chem.MolFromSmiles(s)
    m = Chem.AddHs(m)
    AllChem.EmbedMolecule(m)
    return m

print(read_smile(s).GetConformers()[0].GetPositions())
```

### PR DESCRIPTION
**Pull request by ptosco**
This PR fixes #3492 as well as other issues connected to C++ objects being destroyed too early.
For example, the current RDKit master branch segfaults when doing:

```
In [1]: from rdkit import Chem

In [2]: Chem.MolFromSmiles("CC").GetBonds()[0].GetOwningMol()
Segmentation fault
```

or

```
In [1]: from rdkit import Chem

In [2]: Chem.MolFromSmiles("CC").GetAtoms()[0].GetOwningMol()
Segmentation fault
```

while after this PR this does not happen anymore.

### CODE REVIEW

**Comment by bp-kelley** [comment]
I initially thought that this solution was a pretty have hammer, but it really only changes reference_internal_object to with_custodian_and_ward_postcall and let's conformers along for the ride.

I'm a bit curious on the performance hit, conformers aren't a big deal but atoms may be.  Would this go away if we simply didn't have GetOwningMol?

**Comment by ptosco** [comment]
Benchmark:

```
import rdkit
rdkit.__version__
'2021.03.1dev1'

import gzip
from rdkit import Chem

with gzip.open("/home/toscopa1/sdf/chembl_27.sdf.gz", "rb") as hnd:
    suppl = Chem.ForwardSDMolSupplier(hnd)
    mols = [next(suppl) for i in range(100000)]

<h2><code>master</code> branch</h2>
```
%%timeit
for m in mols:
    for i in range(m.GetNumAtoms()):
        assert(m.GetAtoms()[i].GetOwningMol())
6.82 s ± 151 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

<h2><code>getconformers</code> branch</h2>
```
%%timeit
for m in mols:
    for i in range(m.GetNumAtoms()):
        assert(m.GetAtoms()[i].GetOwningMol())
8.09 s ± 85.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

I'd rather live with a small performance hit than with dangling pointers.
Note that the proposed change involves no copying - only pushing `ROMOL_SPTR`s in the `ReadOnlySeq` iterators and using `with_custodian_and_ward_postcall` to keep `ReadOnlySeq` iterators, and hence the owning mol `ROMOL_SPTR` alive.

I think removing `GetOwningMol()` would break <i>a lot</i> of existing Python code.

**Comment by greglandrum** [comment]
> I think removing GetOwningMol() would break a lot of existing Python code.
Yeah, we can't do that

File: Code/GraphMol/Wrap/Mol.cpp [hunk 1]
```diff
@@ -396,7 +393,11 @@ struct mol_wrapper {
                  1, python::with_custodian_and_ward_postcall<0, 1>>())
 
         .def("GetConformers", GetMolConformers,
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/Mol.cpp:395]: Switching to not returning a tuple here is a clever solution to this problem.

**Comment by bp-kelley** [comment]
I'll leave it to you all, but it is a 15% performance hit for the iterator.  However it does seem to solve this vexing problem

```
atoms = list(mol.GetAtoms())
del mol
# are the atoms valid?  Probably not.
```

So for general safety, I'm for it.  I don't think there is a reasonable way to have GetOwningMol return None as the C++ side thinks the atoms are deleted without some a lot of effort.

**Comment by bp-kelley** [comment]
The only other solution I've seen is to add hidden properties to the output iteratators, like

```
setattr(atom, "_parent", mol)
```

Which probably isn't any faster than the custodian.

**Comment by ptosco** [comment]
@bp-kelley An alternative approach to the `custodian_and_ward` one would be having a Python-accessible, global thread-safe `std::set<ROMol *>` where C++ `ROMol`s are registered upon construction and de-registered upon destruction. Python would check the `std::set<ROMol *>` and throw (or return `None`, depending on the case) when the underlying `ROMol` turns out not to exist anymore. This C++ machinery would be compiled only if Python wrappers are generated. Doable, not sure how much faster than the `custodian_and_ward` one.

**Comment by ptosco** [comment]
@bp-kelley As you say, the `setattr` hack is probably not faster and certainly uglier as it is visible Python-side - I admit that I thought of that at first but then I rejected it as too ugly :-)

**Comment by ptosco** [comment]
@bp-kelley @greglandrum Let me have a play to an alternative solution I have just thought of and see if it is faster. Then we can make a decision on this.

**Comment by greglandrum** [comment]
Please be careful of engaging in premature optimization here. Sure calling `GetAtoms()` is slower, but normally what we do inside the loop that follows that takes considerably longer.

If there's a straightforward way to avoid the performance hit that's great, but given that I don't think this is going to have a dramatic performance impact on real scripts/programs, I don't think it's tragic.

**Comment by greglandrum** [comment]
I find this much cleaner and less disruptive than the alternative in #3530 and as I said in the previous comment, I don't think the real-world performance impact is likely to be that dramatic.

**Review by greglandrum — APPROVED** [review]

**Comment by ptosco** [comment]
@greglandrum Thanks Greg! I'll close #3530. You are right it is more disruptive, but I think it is clean. I'll keep it around for future reference should we ever run into similar lifetime issues that cannot be effectively handled with the simpler `custodian_and_ward` approach.

**Pull request merged**